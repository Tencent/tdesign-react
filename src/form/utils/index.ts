import type { NamePath } from '../type';

// 获取 formMap 管理的数据
export function getMapValue(name: NamePath, formMapRef: React.MutableRefObject<Map<any, any>>) {
  if (!formMapRef.current) return;

  // 提取所有 map key
  const mapKeys = [...formMapRef.current.keys()];
  // 转译为字符串后比对 key 兼容数组格式
  const key = mapKeys.find((key) => String(key) === String(name));
  // 拿到 key 引用地址获取 value
  return formMapRef.current.get(key);
}

// 将数据整理成对象，数组 name 转化嵌套对象: ['user', 'name'] => { user: { name: '' } }
export function calcFieldValue(name: NamePath, value: any) {
  if (Array.isArray(name)) {
    const fieldValue = name.reduceRight((prev, curr) => {
      const arr = [];
      if (/^\d+$/.test(String(curr))) arr[curr] = prev;
      return arr.length ? arr : { [curr]: prev };
    }, value);
    return { ...fieldValue };
  }

  return { [name]: value };
}

// 通过对象数据类型获取 map 引用: { user: { name: '' } } => formMap.get(['user', 'name'])
export function travelMapFromObject(
  obj: Record<any, any>,
  formMapRef: React.MutableRefObject<Map<any, any>>,
  callback: Function,
) {
  for (const [mapName, formItemRef] of formMapRef.current.entries()) {
    // 支持嵌套数据结构
    if (Array.isArray(mapName)) {
      // 创建唯一临时变量 symbol
      const symbol = Symbol('name');
      let fieldValue = null;

      for (let i = 0; i < mapName.length; i++) {
        const item = mapName[i];
        if (Reflect.has(fieldValue || obj, item)) {
          fieldValue = Reflect.get(fieldValue || obj, item);
        } else {
          // 当反射无法获取到值则重置为 symbol
          fieldValue = symbol;
          break;
        }
      }

      // 非 symbol 说明获取到了值
      if (fieldValue !== symbol) {
        callback(formItemRef, fieldValue);
      }
    } else if (Reflect.has(obj, mapName)) {
      callback(formItemRef, obj[mapName]);
    }
  }
}
