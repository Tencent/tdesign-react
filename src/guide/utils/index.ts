import { scrollToParentVisibleArea } from './getScrollParent';
import getRelativePosition from './getRelativePosition';
import getTargetElm from './getTargetElm';
import scrollToElm from './scrollToElm';
import useMount from './useMount';

export { scrollToParentVisibleArea, getRelativePosition, getTargetElm, scrollToElm, useMount };
