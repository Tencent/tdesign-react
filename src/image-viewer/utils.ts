export const downloadFile = function (imgSrc) {
  const image = new Image();
  const name = imgSrc?.split?.('/').pop() || Math.random().toString(32).slice(2);

  image.setAttribute('crossOrigin', 'anonymous');

  image.onload = () => {
    const canvas = document.createElement('canvas');
    canvas.width = image.width;
    canvas.height = image.height;

    const context = canvas.getContext('2d');
    context.drawImage(image, 0, 0, image.width, image.height);
    canvas.toBlob((blob) => {
      const url = URL.createObjectURL(blob);
      const a = document.createElement('a');
      a.download = name;
      a.href = url;
      a.click();
      a.remove();
      URL.revokeObjectURL(url);
    });
  };
  image.src = imgSrc;
};
