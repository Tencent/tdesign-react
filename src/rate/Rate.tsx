import React, { MouseEvent, useState, forwardRef } from 'react';
import classNames from 'classnames';
import { StarFilledIcon as TdStarFilledIcon } from 'tdesign-icons-react';
import { TooltipLite } from '../tooltip';
import { TdRateProps } from './type';
import { StyledProps } from '../common';
import useConfig from '../hooks/useConfig';
import useGlobalIcon from '../hooks/useGlobalIcon';
import useControlled from '../hooks/useControlled';
import { rateDefaultProps } from './defaultProps';

export interface RateProps extends TdRateProps, StyledProps {}

const Rate = forwardRef((props: RateProps, ref: React.Ref<HTMLDivElement>) => {
  const { allowHalf, color, count, disabled, gap, showText, size, texts, icon, className, style, onChange } = props;

  const { classPrefix } = useConfig();
  const { StarFilledIcon } = useGlobalIcon({ StarFilledIcon: TdStarFilledIcon });
  const [starValue = 0, setStarValue] = useControlled(props, 'value', onChange);

  const [hoverValue = undefined, setHoverValue] = useState<number | undefined>(undefined);
  const displayValue = hoverValue || starValue;
  const rootRef = React.useRef(null);

  const activeColor = Array.isArray(color) ? color[0] : color;
  const defaultColor = Array.isArray(color) ? color[1] : 'var(--td-bg-color-component)';

  // 评分图标
  const RateIcon = (props: any) => {
    if (React.isValidElement(icon)) {
      return React.cloneElement(icon, props);
    }
    return <StarFilledIcon {...props} />;
  };

  const getStarValue = (event: MouseEvent<HTMLElement>, index: number) => {
    if (allowHalf) {
      const rootNode = rootRef.current;
      const { left } = rootNode.getBoundingClientRect();
      const firstStar = rootNode.firstChild as HTMLElement;
      const { width } = firstStar.getBoundingClientRect();
      const { clientX } = event;
      const starMiddle = width * (index - 0.5) + gap * (index - 1);
      if (clientX - left >= starMiddle) return index;
      if (clientX - left < starMiddle) return index - 0.5;
    }

    return index;
  };

  const mouseEnterHandler = (event: MouseEvent<HTMLElement>, index: number) => {
    if (disabled) return;
    setHoverValue(getStarValue(event, index));
  };

  const mouseLeaveHandler = () => {
    if (disabled) return;
    setHoverValue(undefined);
  };

  const clickHandler = (event: MouseEvent<HTMLElement>, index: number) => {
    if (disabled) return;
    setStarValue(getStarValue(event, index));
  };

  const getStarCls = (index: number) => {
    if (allowHalf && index + 0.5 === displayValue) return `${classPrefix}-rate__item--half`;
    if (index >= displayValue) return '';
    if (index < displayValue) return `${classPrefix}-rate__item--full`;
  };

  return (
    <div
      ref={ref}
      style={style}
      className={classNames(`${classPrefix}-rate`, className)}
      onMouseLeave={mouseLeaveHandler}
    >
      <ul className={`${classPrefix}-rate__list`} style={{ gap }} ref={rootRef}>
        {[...Array(count)].map((_, index) => (
          <li
            key={index}
            className={classNames(`${classPrefix}-rate__item`, getStarCls(index))}
            onClick={(event) => clickHandler(event, index + 1)}
            onMouseMove={(event) => mouseEnterHandler(event, index + 1)}
          >
            {showText ? (
              <TooltipLite key={index} content={texts[displayValue - 1]}>
                <div className={`${classPrefix}-rate__star-top`}>
                  <RateIcon size={size} color={activeColor} />
                </div>
                <div className={`${classPrefix}-rate__star-bottom`}>
                  <RateIcon size={size} color={defaultColor} />
                </div>
              </TooltipLite>
            ) : (
              <>
                <div className={`${classPrefix}-rate__star-top`}>
                  <RateIcon size={size} color={activeColor} />
                </div>
                <div className={`${classPrefix}-rate__star-bottom`}>
                  <RateIcon size={size} color={defaultColor} />
                </div>
              </>
            )}
          </li>
        ))}
      </ul>
      {showText && <div className={`${classPrefix}-rate__text`}>{texts[displayValue - 1]}</div>}
    </div>
  );
});

Rate.displayName = 'Rate';
Rate.defaultProps = rateDefaultProps;

export default Rate;
