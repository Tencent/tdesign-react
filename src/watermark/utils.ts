const toLowercaseSeparator = (key: string) => key.replace(/([A-Z])/g, '-$1').toLowerCase();

// style对象转字符串
export const getStyleStr = (style: React.CSSProperties): string =>
  Object.keys(style)
    .map((key: keyof React.CSSProperties) => `${toLowercaseSeparator(key)}: ${style[key]};`)
    .join(' ');
