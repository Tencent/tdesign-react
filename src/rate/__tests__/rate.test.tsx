import React from 'react';
import { render, fireEvent, vi } from '@test/utils';
import Rate from '../Rate';

describe('Rate 组件测试', () => {
  // 测试渲染
  test('create', async () => {
    const { container } = render(<Rate />);
    expect(container.firstChild.classList.contains('t-rate')).toBeTruthy();
    expect(document.querySelectorAll('.t-rate__item')).toHaveLength(5);
  });

  // 点击测试
  test('onChange', async () => {
    const clickFn = vi.fn();
    render(<Rate onChange={clickFn} />);
    fireEvent.click(document.querySelector('.t-rate__item'));
    expect(clickFn).toBeCalledTimes(1);
    expect(clickFn).toBeCalledWith(1);
  });

  // mouse 事件测试
  test('mouse', async () => {
    render(<Rate defaultValue={3} showText />);
    expect(document.querySelectorAll('.t-rate__item--full')).toHaveLength(3);

    fireEvent.mouseMove(document.querySelector('.t-rate__item'));
    expect(document.querySelectorAll('.t-rate__item--full')).toHaveLength(1);

    fireEvent.mouseLeave(document.querySelector('.t-rate__item'));
    expect(document.querySelectorAll('.t-rate__item--full')).toHaveLength(3);
  });

  // 数量测试
  test('count', async () => {
    render(<Rate count={10} />);
    expect(document.querySelectorAll('.t-rate__item')).toHaveLength(10);
  });

  describe('props', () => {
    test('disable', async () => {
      const clickFn = vi.fn();
      render(<Rate disabled onChange={clickFn} />);
      fireEvent.click(document.querySelector('.t-rate__item'));
      expect(clickFn).toBeCalledTimes(0);
    });

    test('icon', () => {
      const { container } = render(<Rate icon={<span className="custom-node">TNode</span>} />);
      expect(container.firstChild.classList.contains('t-rate')).toBeTruthy();
      expect(document.querySelectorAll('.custom-node')).toHaveLength(10);
    });

    test('allowHalf', () => {
      render(<Rate allowHalf={true} defaultValue={2.5} />);
      // 默认展示
      expect(document.querySelector('.t-rate__item--half')).toBeTruthy();

      const mockedRect = {
        width: 50, // 默认宽度是50
        height: 0,
        top: 0,
        left: 0,
        bottom: 0,
        right: 0,
      };
      window.Element.prototype.getBoundingClientRect = vi.fn().mockReturnValue(mockedRect);

      // 点击全区域
      fireEvent.mouseMove(document.querySelector('.t-rate__item'), {
        clientX: 30,
      });
      expect(document.querySelector('.t-rate__item--half')).toBeFalsy();

      // 点击半区域
      fireEvent.mouseMove(document.querySelector('.t-rate__item'), {
        clientX: 5,
      });
      expect(document.querySelector('.t-rate__item--half')).toBeTruthy();
    });
  });
});
