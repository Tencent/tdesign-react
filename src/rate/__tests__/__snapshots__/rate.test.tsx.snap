// Vitest Snapshot v1

exports[`Rate 组件测试 > count 1`] = `
<div>
  <div
    class="t-rate"
  >
    <ul
      class="t-rate__list"
      style="gap: 4px;"
    >
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Rate 组件测试 > create 1`] = `
<div>
  <div
    class="t-rate"
  >
    <ul
      class="t-rate__list"
      style="gap: 4px;"
    >
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
      <li
        class="t-rate__item"
      >
        <div
          class="t-rate__star-top"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="#ED7B2F"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
        <div
          class="t-rate__star-bottom"
        >
          <svg
            class="t-icon t-icon-star-filled"
            color="var(--td-bg-color-component)"
            fill="none"
            height="1em"
            style="font-size: 24px;"
            viewBox="0 0 16 16"
            width="1em"
          >
            <path
              d="M7.6 1.82a.45.45 0 01.8 0l1.8 3.65 4.03.58c.37.05.52.5.25.77l-2.91 2.84.69 4a.45.45 0 01-.66.48L8 12.25l-3.6 1.9a.45.45 0 01-.65-.48l.68-4.01-2.9-2.84a.45.45 0 01.24-.77l4.03-.58 1.8-3.65z"
              fill="currentColor"
              fill-opacity="0.9"
            />
          </svg>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
